### PR TITLE
client: set timeout for each request

### PR DESCRIPTION
**Pull request by yichengq**
Add Timeout field in Config, so users could set timeout for each request.
Before this, one hanged request may block the call for long time. After
this, if the network is good, the user could set short timeout and expect
that API call can attempt next available endpoint quickly.

from #2850  I borrow a lot of code of the original PR.

### CODE REVIEW

File: client/client_test.go [hunk 1]
```diff
@@ -340,7 +340,7 @@ func TestHTTPClusterClientDo(t *testing.T) {
 		// context.DeadlineExceeded short-circuits Do
 		{
 			client: &httpClusterClient{
-				endpoints: []url.URL{fakeURL, fakeURL},
+				endpoints: []url.URL{fakeURL},
```
Comment (yichengq) [inline client/client_test.go]: The previous test is not correct after this commit. If it has two endpoints, we should expect it try the second one before returning error. Use single endpoint to fix the test.

File: client/client.go [hunk 2]
```diff
@@ -87,6 +87,18 @@ type Config struct {
...
+	// Client. The timeout includes connection time, any
+	// redirects, and reading the response body. If the timer is over
+	// when reading of the Response.Body, the request is counted as failure.
+	//
+	// One API call may include multiple HTTP requests sent to different
+	// endpoints. This Timeout restricts the duration for each single
+	// HTTP request, not the duration of the API call.
+	//
+	// A Timeout of zero means no timeout.
+	Timeout time.Duration
```
Comment (xiang90) [inline client/client.go]: TimeoutPerRequest? I hope we can avoid mentioning HTTP at client level.

Comment (yichengq) [inline client/client.go]: ok. let me try

**Comment by yichengq** [comment]
updated.

File: client/client.go [hunk 3]
```diff
@@ -87,6 +87,17 @@ type Config struct {
 	// Password is the password for the specified user to add as an authorization header
 	// to the request.
 	Password string
+
+	// TimeoutPerRequest specifies a time limit for single request made by this
+	// Client. The timeout includes connection time, any
+	// redirects, and reading the response body.
+	//
+	// One API call may include multiple requests sent to different
```
Comment (xiang90) [inline client/client.go]: One short sentence to explain why it may send to different endpoints?

Comment (yichengq) [inline client/client.go]: added

File: client/client_test.go [hunk 4]
```diff
@@ -337,21 +337,6 @@ func TestHTTPClusterClientDo(t *testing.T) {
 			wantPinned: 1,
 		},
 
-		// context.DeadlineExceeded short-circuits Do
```
Comment (yichengq) [inline client/client_test.go]: DeadlineExceeded cannot short-circuit Do anymore, and original code is not the correct way to test DeadlineExceeded return.

Comment (xiang90) [inline client/client_test.go]: Can we add a deadline reached test to ensure we return the deadline error correctly (not cluster error)

**Comment by xiang90** [comment]
LGTM

**Comment by xiang90** [comment]
@yichengq After reviewing this again, I have a concern about watch request. A global timeout will affect all requests including watches.

**Comment by yichengq** [comment]
updated. PTAL

File: client/client.go [hunk 5]
```diff
@@ -87,6 +87,24 @@ type Config struct {
...
+	//
+	// A TimeoutPerRequest of zero means no timeout.
+	TimeoutPerRequest time.Duration
+
+	// HandshakeTimeoutPerWatchRequest specifies the time limit for handshake
+	// phase of a single watch request. The timeout includes connection time
+	// and acknowledge from server to indicate that watch is under way.
+	//
+	// A HandshakeTimeoutPerWatchRequest of zero means no timeout.
+	HandshakeTimeoutPerWatchRequest time.Duration
```
Comment (xiang90) [inline client/client.go]: I still think a headerTimeoutPerRequest that will apply to all requests would be good enough.

Comment (xiang90) [inline client/client.go]: That will greatly simplify this commit and make everything consistent and easy to understand.

Comment (yichengq) [inline client/client.go]: My only concern is that PUT request needs to commit message before responding, which may take longer time. Will try.

Comment (xiang90) [inline client/client.go]: Will you set today's handshake timeout at sub second level? If not, I would not worry about that. And we can simple doc the fact as well.

Comment (yichengq) [inline client/client.go]: OK. i will doc it then.

File: client/client_test.go (hunk 4 above)
Comment (yichengq) [inline client/client_test.go]: added

**Comment by yichengq** [comment]
updated to use HeaderTimeoutPerRequest and add comment to explain server behavior on different kinds of requests. /cc @jedsmith

File: client/client.go [hunk 6]
```diff
@@ -87,6 +87,23 @@ type Config struct {
 	// Password is the password for the specified user to add as an authorization header
 	// to the request.
 	Password string
+
+	// HeaderTimeoutPerRequest specifies the time limit for to wait for response
+	// header in a single request made by the Client. The timeout includes
+	// connection time, any redirects, and header wait time.
+	//
+	// For non-watch GET request, server returns the response immediately.
```
Comment (xiang90) [inline client/client.go]: returns the response body

File: client/client.go [hunk 7]
```diff
@@ -87,6 +87,23 @@ type Config struct {
 	// Password is the password for the specified user to add as an authorization header
 	// to the request.
 	Password string
+
+	// HeaderTimeoutPerRequest specifies the time limit for to wait for response
```
Comment (xiang90) [inline client/client.go]: time limit for waiting

File: client/client.go (hunk 6 above)
Comment (yichengq) [inline client/client.go]: fixed.

**Comment by xiang90** [comment]
LGTM

**Comment by yichengq** [comment]
:ship:

**Pull request merged**